assets/js/data/cart/push-changes.ts

```diff
@@ -0,0 +1,149 @@
+/**
+ * External dependencies
+ */
+import { debounce } from 'lodash';
+import { select, dispatch } from '@wordpress/data';
+import {
+	formatStoreApiErrorMessage,
+	pluckAddress,
+	pluckEmail,
+} from '@woocommerce/base-utils';
+import {
+	CartResponseBillingAddress,
+	CartResponseShippingAddress,
+} from '@woocommerce/type-defs/cart-response';
+import isShallowEqual from '@wordpress/is-shallow-equal';
+import { BillingAddressShippingAddress } from '@woocommerce/type-defs/cart';
+
+/**
+ * Internal dependencies
+ */
+import { STORE_KEY } from './constants';
+
+declare type CustomerData = {
+	billingData: CartResponseBillingAddress;
+	shippingAddress: CartResponseShippingAddress;
+} | null;
+
+/**
+ * Checks if a cart response contains an email property.
+ */
+const isCartResponseBillingAddress = (
+	address: CartResponseBillingAddress | CartResponseShippingAddress
+): address is CartResponseBillingAddress => {
+	return 'email' in address;
+};
+
+/**
+ * Does a shallow compare of important address data to determine if the cart needs updating on the server. This takes
+ * the current and previous address into account, as well as the billing email field.
+ */
+const isAddressDirty = <
+	T extends CartResponseBillingAddress | CartResponseShippingAddress
+>(
+	// An object containing all previous address information
+	previousAddress: T,
+	// An object containing all address information.
+	address: T
+): boolean => {
+	if (
+		isCartResponseBillingAddress( address ) &&
+		pluckEmail( address ) !==
+			pluckEmail( previousAddress as CartResponseBillingAddress )
+	) {
+		return true;
+	}
+
+	return (
+		!! address.country &&
+		! isShallowEqual(
+			pluckAddress( previousAddress ),
+			pluckAddress( address )
+		)
+	);
+};
+
+/**
+ * Local cache of customerData used for comparisons.
+ */
+let customerData = <CustomerData>null;
+
+/**
+ * Tracks which props have changed so the correct data gets pushed to the server.
+ */
+const dirtyProps = {
+	billingData: false,
+	shippingAddress: false,
+};
+
+/**
+ * Function to dispatch an update to the server. This is debounced.
+ */
+const updateCustomerData = debounce( (): void => {
+	const customerDataToUpdate = {} as Partial< BillingAddressShippingAddress >;
+
+	if ( dirtyProps.billingData && customerData?.billingData ) {
```

Comment:
would there ever be a missing billingData under customerData? Won't that be equal to the default address? Or is this because of customerData being null? Could that be treated separately?

Reply:
This is because of the initial `null` value for customerData and is needed to appease TypeScript. The `null` is needed so we know to populate the customer address from `null` to the received value on first mount.

Reply:
@mikejolley I recommend we add a comment with 
`The null is needed so we know to populate the customer address from null to the received value on first mount.`,
maybe above the declaration of customerData?

Also, I am thinking we should call `updateCustomerData();` only when we have dirty props.
 ```
if (dirtyProps.billingData || dirtyProps.shippingData) {
updateCustomerData();
}
```

and so we can be more intentional in the code.

```
		const { billingData, shippingAddress } = customerData;
		if ( dirtyProps.billingData ) {
			customerDataToUpdate.billing_address = billingData;
			dirtyProps.billingData = false;
		}

		if ( dirtyProps.shippingAddress ) {
			customerDataToUpdate.shipping_address = shippingAddress;
			dirtyProps.shippingAddress = false;
		}
```

Reply:
Fixed in [Tweak customerData to avoid null](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/5817/commits/f3109acb2cf90c5dee5889d41a8f6caadd516e1e)

Reply:
Way better 👏